grr/server/grr_response_server/bin/grrafana.py

```diff
@@ -0,0 +1,185 @@
+      return getattr(self, "On" + endpoint)(request, **values)
+    except werkzeug_exceptions.HTTPException as e:
+      return e
+
+  def WsgiApp(self, environ, start_response) -> werkzeug_wrappers.Response:
+    request = JSONRequest(environ)
+    response = self.DispatchRequest(request)
+    return response(environ, start_response)
+
+  def __call__(self, environ, start_response):
+    return self.WsgiApp(environ, start_response)
+
+  def OnRoot(self, request: JSONRequest) -> JSONResponse:
+    """Returns OK message to database connection check."""
+    return JSONResponse()
+
+  def OnSearch(self, request: JSONRequest) -> JSONResponse:
+    """Depending on the type of request Grafana is issuing, this method returns either
+    available client resource usage metrics from the constant AVAILABLE_METRICS, or
+    possible values for a defined Grafana variable (currently supports only variables based
+    on client IDs)."""
+    if "type" in request.json:
+      # Grafana request issued on Panel > Queries page. Grafana expects the list of metrics
+      # that can be listed on the dropdown-menu called "Metric".
+      response = AVAILABLE_METRICS
+    else:
+      # Grafana request issued on Variables > New/Edit page for variables of type query.
+      # Grafana expectes the list of possible values of the variable.
+      # At the moment, GRRafana doesn't support such variables, so it returns no possible values.
+      response = []
+    return JSONResponse(response=json.dumps(response), mimetype=JSON_MIME_TYPE)
+
+  def OnQuery(self, request: JSONRequest) -> JSONResponse:
+    """Given a client ID as a Grafana variable and targets (resource usages),
+    returns datapoints in a format Grafana can interpret."""
+    json_data = request.json
+    requested_client_id = _ExtractClientIdFromVariable(
+        json_data)  # There must be a ClientID variable declated in Grafana.
+    requested_targets = [entry["target"] for entry in json_data["targets"]]
+    response = _FetchDatapointsForTargets(requested_client_id,
+                                          json_data["maxDataPoints"],
+                                          requested_targets)
+    return JSONResponse(response=json.dumps(response), mimetype=JSON_MIME_TYPE)
+
+  def OnAnnotations(self, request: JSONRequest) -> JSONResponse:
+    pass
+
+
+Datapoint = Tuple[float, int]
+Datapoints = List[Datapoint]
+TargetWithDatapoints = Dict[Text, Datapoints]
+
+
+def _FetchDatapointsForTargets(
+    client_id: Text, limit: int,
+    targets: List[Text]) -> List[TargetWithDatapoints]:
+  """Fetches a list of <datapoint, timestamp> tuples for each target metric from Fleetspeak database."""
+  records_list = fleetspeak_utils.FetchClientResourceUsageRecordsFromFleetspeak(
+      client_id, limit)
+  response = []
+  for target in targets:
+    datapoints_for_single_target = _CreateDatapointsForTarget(
+        target, records_list)
+    target_datapoints_dict = cast(TargetWithDatapoints, {
```

Comment:
The type definition is unfitting: `TargetWithDatapoints` is defined as Dict with string keys and Datapoints values, yet you assign "target" to a string. Dict is not the type you want here, define a class or a namedtuple instead (our Python version does not support dataclasses yet).

Reply:
Then, this cast is no longer needed.

Reply:
I changed `TargetWithDatapoints` to a namedtuple like you suggested. The only caveat is that Grafana cannot interpret correctly the JSON that is extracted from this namedtuple, so I use [`as_dict()`](https://docs.python.org/3/library/collections.html#collections.somenamedtuple._asdict) in the method `_OnQuery`.

Let me know if you think this is the right way to go about this, because it feels like a slight workaround.